src/compiler/checker.ts

```diff
@@ -26465,10 +26465,10 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                     // and has inferences that would conflict. Otherwise, we prefer the contra-variant inference.
                     const preferCovariantType = inferredCovariantType && (!inferredContravariantType ||
                         !(inferredCovariantType.flags & TypeFlags.Never) &&
-                            some(inference.contraCandidates, t => isTypeSubtypeOf(inferredCovariantType, t)) &&
+                            some(inference.contraCandidates, t => isTypeAssignableTo(inferredCovariantType, t)) &&

```

Comment:
If this is not the correct change then at the very least a test case should be proposed that shows how the `isTypeSubtypeOf` is better.

When this check was originally introduced by @ahejlsberg [here](https://github.com/microsoft/TypeScript/pull/27028#discussion_r217112232) it was states that:
> Furthermore, knowing that an error will result when the co-variant inference is not a subtype of the contra-variant inference, we now prefer the contra-variant inference because it is likely to have come from an explicit type annotation on a function. This improves our error reporting.

An alternative idea to fix this example would be to keep `coAndContraRelationCheck` on `inferenceContext`. From what I understand, the `subtypeRelation` in this context is mainly used when dealing with multiple overloads - a single signature case uses `assignableRelation` with `chooseOverload`. So perhaps the used relation should determine what is being used here.